snapcraft/internal/build_providers/_base_provider.py

```diff
@@ -0,0 +1,97 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import abc
+import shlex
+from typing import Callable
+
+import petname
+
+
+class Provider():
```

Comment:
This is heading the right direction, but I'm losing track of what is and isn't public API ("public" defined to be "meant to be called by a user of `Provider`). For example, `run` and `launch` are both public, but I don't believe they're actually meant to be called by anything except `Provider`, right?